File: apps/web/src/stores/ticket.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/lib/api'
import type { Status } from './status'

export interface TicketServiceSnapshot {
  id: string
  service_id: string | null
  captured_name: string
  captured_price_mxn: number | null
  captured_duration_minutes: number
  sort_order: number
}

export interface Ticket {
  id: string
  organization_id: string
  location_id: string
  public_code: string
  plate: string | null
  vehicle_desc: string | null
  manual_ticket_no: string | null
  customer_name: string | null
  customer_whatsapp: string | null
  current_status_id: string
  current_status: Status
  eta_at: string | null
  created_at: string
  updated_at: string
  completed_at: string | null
  services: TicketServiceSnapshot[]
  total_duration_minutes: number
  total_price_mxn: number | null
  package_snapshot: PackageSnapshot | null
}

export interface TicketListItem {
  id: string
  public_code: string
  plate: string | null
  vehicle_desc: string | null
  current_status_id: string
  current_status: Status
  eta_at: string | null
  created_at: string
  services_summary: string
  total_duration_minutes: number
}

export interface TicketListResponse {
  items: TicketListItem[]
  total: number
}

export interface PackageSnapshot {
  package_id: string
  name: string
  price_mxn: number
  workers_required: number
  estimated_duration_minutes: number
}

export interface TicketCreate {
  plate?: string | null
  vehicle_desc?: string | null
  manual_ticket_no?: string | null
  customer_name?: string | null
  customer_whatsapp?: string | null
  service_ids?: string[]
  package_id?: string | null
  vehicle_type_id?: string | null
}

export interface TicketUpdate {
  plate?: string | null
  vehicle_desc?: string | null
  manual_ticket_no?: string | null
  customer_name?: string | null
  customer_whatsapp?: string | null
  service_ids?: string[]
}

export interface TicketEvent {
  id: string
  ticket_id: string
  actor_user_id: string
  actor_name: string | null
  event_type: string
  from_status_id: string | null
  from_status_name: string | null
  to_status_id: string | null
  to_status_name: string | null
  payload_json: Record<string, unknown> | null
  happened_at: string
}

export interface TicketEventListResponse {
  items: TicketEvent[]
  total: number
}

export interface TicketFilters {
  status_id?: string
  q?: string
  date?: 'today' | null
}

export const useTicketStore = defineStore('ticket', () => {
  const tickets = ref<TicketListItem[]>([])
  const currentTicket = ref<Ticket | null>(null)
  const events = ref<TicketEvent[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)
  const total = ref(0)

  async function listTickets(locationId: string, filters: TicketFilters = {}) {
    loading.value = true
    error.value = null
    try {
      const params: Record<string, string> = {}
      if (filters.status_id) params.status_id = filters.status_id
      if (filters.q) params.q = filters.q
      if (filters.date) params.date = filters.date

      const response = await api.get<TicketListResponse>(
        `/api/v1/locations/${locationId}/tickets`,
        { params }
      )
      tickets.value = response.data.items
      total.value = response.data.total
    } catch (err) {
      error.value = 'Error al cargar tickets'
      console.error('Failed to fetch tickets:', err)
    } finally {
      loading.value = false
    }
  }

  async function getTicket(id: string) {
    loading.value = true
    error.value = null
    try {
      const response = await api.get<Ticket>(`/api/v1/tickets/${id}`)
      currentTicket.value = response.data
      return response.data
    } catch (err) {
      error.value = 'Error al cargar ticket'
      console.error('Failed to fetch ticket:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  async function createTicket(locationId: string, data: TicketCreate) {
    const response = await api.post<Ticket>(
      `/api/v1/locations/${locationId}/tickets`,
      data
    )
    return response.data
  }

  async function updateTicket(id: string, data: TicketUpdate) {
    const response = await api.patch<Ticket>(`/api/v1/tickets/${id}`, data)
    currentTicket.value = response.data
    return response.data
  }

  async function moveTicket(id: string, toStatusId: string) {
    const response = await api.post<Ticket>(`/api/v1/tickets/${id}/move`, {
      to_status_id: toStatusId,
    })
    currentTicket.value = response.data
    return response.data
  }

  async function updateEta(id: string, etaAt: string | null) {
    const response = await api.patch<Ticket>(`/api/v1/tickets/${id}/eta`, {
      eta_at: etaAt,
    })
    currentTicket.value = response.data
    return response.data
  }

  async function listEvents(ticketId: string) {
    const response = await api.get<TicketEventListResponse>(
      `/api/v1/tickets/${ticketId}/events`
    )
    events.value = response.data.items
    return response.data.items
  }

  function clearCurrentTicket() {
    currentTicket.value = null
    events.value = []
  }

  return {
    tickets,
    currentTicket,
    events,
    loading,
    error,
    total,
    listTickets,
    getTicket,
    createTicket,
    updateTicket,
    moveTicket,
    updateEta,
    listEvents,
    clearCurrentTicket,
  }
})

File: apps/api/src/washflow_api/schemas/ticket.py
"""Ticket schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator

from washflow_api.schemas.status import StatusOut


class TicketServiceSnapshot(BaseModel):
    """Snapshot of a service attached to a ticket."""

    id: str
    service_id: str | None
    captured_name: str
    captured_price_mxn: int | None
    captured_duration_minutes: int
    sort_order: int

    model_config = {"from_attributes": True}


class TicketCreate(BaseModel):
    """Ticket creation schema."""

    plate: str | None = Field(default=None, max_length=16)
    vehicle_desc: str | None = Field(default=None, max_length=80)
    manual_ticket_no: str | None = Field(default=None, max_length=24)
    customer_name: str | None = Field(default=None, max_length=60)
    customer_whatsapp: str | None = Field(default=None, max_length=20)

    # Package-based creation (preferred)
    package_id: str | None = None
    vehicle_type_id: str | None = None

    # Legacy: manual service selection (for backwards compatibility)
    service_ids: list[str] = Field(default_factory=list)

    @field_validator("plate", "vehicle_desc", "manual_ticket_no", "customer_name")
    @classmethod
    def strip_string(cls, v: str | None) -> str | None:
        if v is not None:
            v = v.strip()
            return v if v else None
        return v

    @field_validator("customer_whatsapp")
    @classmethod
    def validate_whatsapp(cls, v: str | None) -> str | None:
        if v is not None:
            v = v.strip()
            if not v:
                return None
            # Basic validation: allow +52 prefix or 10 digits
            cleaned = v.replace(" ", "").replace("-", "")
            if not cleaned:
                return None
            # Accept +52... or just digits
            if cleaned.startswith("+"):
                if len(cleaned) < 10:
                    raise ValueError("Número de WhatsApp inválido")
            else:
                if not cleaned.isdigit():
                    raise ValueError("Número de WhatsApp debe contener solo dígitos")
                if len(cleaned) < 10:
                    raise ValueError("Número de WhatsApp inválido")
            return v
        return v


class TicketUpdate(BaseModel):
    """Ticket update schema."""

    plate: str | None = Field(default=None, max_length=16)
    vehicle_desc: str | None = Field(default=None, max_length=80)
    manual_ticket_no: str | None = Field(default=None, max_length=24)
    customer_name: str | None = Field(default=None, max_length=60)
    customer_whatsapp: str | None = Field(default=None, max_length=20)
    service_ids: list[str] | None = None

    @field_validator("plate", "vehicle_desc", "manual_ticket_no", "customer_name")
    @classmethod
    def strip_string(cls, v: str | None) -> str | None:
        if v is not None:
            v = v.strip()
            return v if v else None
        return v

    @field_validator("customer_whatsapp")
    @classmethod
    def validate_whatsapp(cls, v: str | None) -> str | None:
        if v is not None:
            v = v.strip()
            if not v:
                return None
            cleaned = v.replace(" ", "").replace("-", "")
            if not cleaned:
                return None
            if cleaned.startswith("+"):
                if len(cleaned) < 10:
                    raise ValueError("Número de WhatsApp inválido")
            else:
                if not cleaned.isdigit():
                    raise ValueError("Número de WhatsApp debe contener solo dígitos")
                if len(cleaned) < 10:
                    raise ValueError("Número de WhatsApp inválido")
            return v
        return v


class PackageSnapshot(BaseModel):
    """Snapshot of package info captured at ticket creation."""

    id: str | None
    name: str | None
    price_mxn: int | None
    workers_required: int | None
    estimated_duration_minutes: int | None
    vehicle_type_id: str | None
    vehicle_type_name: str | None

    model_config = {"from_attributes": True}


class TicketOut(BaseModel):
    """Ticket response schema."""

    id: str
    organization_id: str
    location_id: str
    public_code: str
    plate: str | None
    vehicle_desc: str | None
    manual_ticket_no: str | None
    customer_name: str | None
    customer_whatsapp: str | None
    current_status_id: str
    current_status: StatusOut
    eta_at: datetime | None
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None
    services: list[TicketServiceSnapshot]
    total_duration_minutes: int
    total_price_mxn: int | None

    # Package info
    package_id: str | None = None
    vehicle_type_id: str | None = None
    package_snapshot: PackageSnapshot | None = None

    model_config = {"from_attributes": True}


class TicketListItem(BaseModel):
    """Ticket list item schema (lighter than full TicketOut)."""

    id: str
    public_code: str
    plate: str | None
    vehicle_desc: str | None
    current_status_id: str
    current_status: StatusOut
    eta_at: datetime | None
    created_at: datetime
    services_summary: str  # e.g. "Lavado completo, Encerado"
    total_duration_minutes: int

    model_config = {"from_attributes": True}


class TicketListResponse(BaseModel):
    """Paginated ticket list response."""

    items: list[TicketListItem]
    total: int


class TicketMoveRequest(BaseModel):
    """Request to move ticket to another status."""

    to_status_id: str


class EtaUpdateRequest(BaseModel):
    """Request to update ticket ETA."""

    eta_at: datetime | None


class TicketEventOut(BaseModel):
    """Ticket event response schema."""

    id: str
    ticket_id: str
    actor_user_id: str
    actor_name: str | None = None
    event_type: str
    from_status_id: str | None
    from_status_name: str | None = None
    to_status_id: str | None
    to_status_name: str | None = None
    payload_json: dict[str, Any] | None
    happened_at: datetime

    model_config = {"from_attributes": True}


class TicketEventListResponse(BaseModel):
    """List of ticket events."""

    items: list[TicketEventOut]
    total: int


class PublicTicketOut(BaseModel):
    """Public ticket response schema (minimal, non-sensitive data only)."""

    public_code: str
    vehicle_label: str  # plate or vehicle_desc, computed
    service_name: str | None  # e.g. "Lavado completo" or "Lavado completo, Encerado"
    status_name: str
    status_is_terminal: bool
    eta_at: datetime | None
    updated_at: datetime
    location_name: str
    organization_name: str

    # Branding fields (resolved with inheritance: location override > org > fallback)
    brand_name: str  # Display name for branding
    brand_logo_url: str | None  # URL to logo image
    brand_primary_color: str | None  # HEX color like #2563eb

    model_config = {"from_attributes": True}
